lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs

```diff
@@ -282,38 +588,143 @@ public void OnCompleted()
+        }
+
+        private string GetMapperTaskIdByEvaluatorId(string evaluatorId)
         {
             return string.Format("{0}-{1}-Version0",
                 IMRUConstants.MapTaskPrefix,
                 _serviceAndContextConfigurationProvider.GetPartitionIdByEvaluatorId(evaluatorId));
         }
 
+        private void DoneAction()
+        {
+            ShutDownAllEvaluators();
+        }
+
+        private void FailAction()
+        {
+            ShutDownAllEvaluators();
+        }
+
         /// <summary>
-        /// Shuts down evaluators once all completed task messages are received
+        /// Shuts down evaluators
         /// </summary>
         private void ShutDownAllEvaluators()
         {
-            foreach (var task in _completedTasks)
+            foreach (var context in _contextManager.ActiveContexts)
+            {
+                Logger.Log(Level.Verbose, string.Format(CultureInfo.InvariantCulture, "Disposing active context: {0}", context.Id));
+                context.Dispose();
+            }
+        }
+
+        /// <summary>
+        /// Resets Failed Evaluators
+        /// Changes state to WaitingForEvaluator if it is in recovery.
+        /// If it is first time, request both master and map evaluators
+        /// If it is in recovery, based on the count of failed evaluators, requests missing map evaluators
+        /// </summary>
+        private void StartAction()
+        {
+            lock (_lock)
             {
-                Logger.Log(Level.Info, string.Format("Disposing task: {0}", task.Id));
-                task.ActiveContext.Dispose();
+                if (!_isFirstTry)
+                {
+                    _numberOfRetryForFaultTolerant++;
+                    Logger.Log(Level.Info, string.Format("Start recovery with _numberOfRetryForFaultTolerant:" + _numberOfRetryForFaultTolerant));
+                }
+
+                if (_isFirstTry)
+                {
+                    _systemState = new SystemStateMachine();
+                }
+                else
+                {
+                    _systemState.MoveNext(SystemStateEvent.Recover);
+                }
+
+                var mappersToRequest = _isFirstTry ? _totalMappers : _evaluatorManager.NumberofFailedMappers();
+                _evaluatorManager.ResetFailedEvaluators();
+
+                if (mappersToRequest == 0)
+                {
+                    Logger.Log(Level.Info, "There is no failed Evaluator in this recovery.");
+                    if (_contextManager.AreAllContextsReceived)
+                    {
+                        OnNext(_contextManager.NumberOfActiveContexts);
+                    }
+                    else
+                    {
+                        Exceptions.Throw(new IMRUSystemException("In recovery, there is no Failed evaluator but not all the contexts are received"), Logger);
+                    }
+                    return;
+                }
+
+                if (_isFirstTry)
```

Comment:
Previously we need to recover both master and mapper evaluators. So first time and later recovery have something in common to share in terns of requesting master and mapper Evaluators. 
Since we are not going to recover master evaluator now, I would just move the code for first time inside IDriverStarted handler, leaving recovery in this method, that would make the code easy to read.